pkg/stanza/fileconsumer/header.go

```diff
@@ -0,0 +1,275 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fileconsumer // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer"
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+
+	"go.uber.org/zap"
+	"golang.org/x/text/encoding"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/entry"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/pipeline"
+)
+
+type HeaderConfig struct {
+	Pattern           string            `mapstructure:"pattern"`
+	MetadataOperators []operator.Config `mapstructure:"metadata_operators"`
+
+	// these are set by the "build" function
+	matchRegex *regexp.Regexp
+	splitFunc  bufio.SplitFunc
+}
+
+// validate returns an error describing why the configuration is invalid, or nil if the configuration is valid.
+func (hc *HeaderConfig) validate() error {
+	_, err := regexp.Compile(hc.Pattern)
+	if err != nil {
+		return fmt.Errorf("invalid `pattern`: %w", err)
+	}
+
+	if len(hc.MetadataOperators) == 0 {
+		return errors.New("at least one operator must be specified for `metadata_operators`")
+	}
+
+	nopLogger := zap.NewNop().Sugar()
+	outOp := newHeaderPipelineOutput(nopLogger)
+	p, err := pipeline.Config{
+		Operators:     hc.MetadataOperators,
+		DefaultOutput: outOp,
+	}.Build(nopLogger)
+
+	if err != nil {
+		return fmt.Errorf("failed to build pipelines: %w", err)
+	}
+
+	if !p.Operators()[0].CanProcess() {
+		return errors.New("first operator must be able to process entries")
+	}
+
+	return nil
+}
+
+func (hc *HeaderConfig) build(enc encoding.Encoding) error {
+	var err error
+	hc.matchRegex, err = regexp.Compile(hc.Pattern)
+	if err != nil {
+		return fmt.Errorf("failed to compile multiline pattern: %w", err)
+	}
+
+	hc.splitFunc, err = helper.NewNewlineSplitFunc(enc, false, func(b []byte) []byte {
+		return bytes.Trim(b, "\r\n")
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create split func: %w", err)
+	}
+
+	return nil
+}
+
+// buildHeader builds a header struct from the header config.
+func (hc *HeaderConfig) buildHeader(logger *zap.SugaredLogger, persister operator.Persister) (*header, error) {
+	outOp := newHeaderPipelineOutput(logger)
+	p, err := pipeline.Config{
+		Operators:     hc.MetadataOperators,
+		DefaultOutput: outOp,
+	}.Build(logger)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to build pipeline: %w", err)
+	}
+
+	scopedPersister := operator.NewScopedPersister("header", persister)
+
+	if err := p.Start(scopedPersister); err != nil {
```

Comment:
I am wondering if it's really worth trying to persist anything before we get through the header. What if we just pass in a `storage.NewNopClient()` here instead? Of course we'd need to ensure that the file is also not checkpointed until we are through the header, so that in the event of a restart, we'd start from scratch. WDYT?

Reply:
I agree, it makes sense to pass a NopClient.

Currently, the header doesn't checkpoint (it's re-read everytime) - I can try and implement checkpointing in this PR, or could do that in a follow-up PR if that would make it easier to review

Reply:
Let's leave it out for now. We can consider it in the future.